File: processor/src/main/java/com/github/snowdream/toybricks/processor/ProcessorManager.java
package com.github.snowdream.toybricks.processor;

import javax.annotation.processing.Filer;
import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import javax.tools.Diagnostic;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by snowdream on 17/2/12.
 */
public class ProcessorManager implements Handler {
    private ProcessingEnvironment mProcessingEnvironment;
    private List<BaseContainerHandler> mHandlers;

    private Types mTypeUtils;
    private Elements mElementUtils;
    private Filer mFiler;
    private Messager mMessager;


    ProcessorManager(ProcessingEnvironment processingEnv) {
        mProcessingEnvironment = processingEnv;

        mTypeUtils = processingEnv.getTypeUtils();
        mElementUtils = processingEnv.getElementUtils();
        mFiler = processingEnv.getFiler();
        mMessager = processingEnv.getMessager();
    }

    void addHandlers(BaseContainerHandler... containerHandlers) {
        if (containerHandlers == null || containerHandlers.length <= 0) return;

        for (BaseContainerHandler handler :
                containerHandlers) {

            if (mHandlers == null) {
                mHandlers = new ArrayList<>();
            }

            if (!mHandlers.contains(handler)) {
                mHandlers.add(handler);
            }
        }
    }

    @Override
    public void handle(ProcessorManager processorManager, RoundEnvironment roundEnvironment) {
        if (mHandlers != null){
            for (BaseContainerHandler handler:
                    mHandlers ) {
                handler.handle(processorManager,roundEnvironment);
            }
        }

    }

    /**
     * Prints an error message
     *
     * @param e The element which has caused the error. Can be null
     * @param msg The error message
     */
    public void error(Element e, String msg) {
        mMessager.printMessage(Diagnostic.Kind.ERROR, msg, e);
    }

    /**
     * Prints an info message
     *
     * @param e The element which has caused the error. Can be null
     * @param msg The error message
     */
    public void info(Element e, String msg) {
        mMessager.printMessage(Diagnostic.Kind.NOTE, msg, e);
    }

    public Types getTypeUtils(){
        return mTypeUtils;
    }

    public Elements getElementUtils(){
        return mElementUtils;
    }

    public Filer getFiler(){
        return mFiler;
    }

    public Messager getMessager(){
        return mMessager;
    }
}

File: processor/src/main/kotlin/com/github/snowdream/toybricks/processor/ProcessorManager.kt
package com.github.snowdream.toybricks.processor

import javax.annotation.processing.Filer
import javax.annotation.processing.Messager
import javax.annotation.processing.ProcessingEnvironment
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.Element
import javax.lang.model.util.Elements
import javax.lang.model.util.Types
import javax.tools.Diagnostic
import java.util.ArrayList

/**
 * Created by snowdream on 17/2/12.
 */
class ProcessorManager internal constructor(private val mProcessingEnvironment: ProcessingEnvironment) : Handler {
    private var mHandlers: MutableList<BaseContainerHandler>? = null

    val typeUtils: Types
    val elementUtils: Elements
    val filer: Filer
    val messager: Messager


    init {

        typeUtils = mProcessingEnvironment.typeUtils
        elementUtils = mProcessingEnvironment.elementUtils
        filer = mProcessingEnvironment.filer
        messager = mProcessingEnvironment.messager
    }

    internal fun addHandlers(vararg containerHandlers: BaseContainerHandler) {
        if (containerHandlers == null || containerHandlers.size <= 0) return

        for (handler in containerHandlers) {

            if (mHandlers == null) {
                mHandlers = ArrayList<BaseContainerHandler>()
            }

            if (!mHandlers!!.contains(handler)) {
                mHandlers!!.add(handler)
            }
        }
    }

    override fun handle(processorManager: ProcessorManager, roundEnvironment: RoundEnvironment) {
        if (mHandlers != null) {
            for (handler in mHandlers!!) {
                handler.handle(processorManager, roundEnvironment)
            }
        }

    }

    /**
     * Prints an error message

     * @param e The element which has caused the error. Can be null
     * *
     * @param msg The error message
     */
    fun error(e: Element?, msg: String) {
        messager.printMessage(Diagnostic.Kind.ERROR, msg, e)
    }

    /**
     * Prints an info message

     * @param e The element which has caused the error. Can be null
     * *
     * @param msg The error message
     */
    fun info(e: Element, msg: String) {
        messager.printMessage(Diagnostic.Kind.NOTE, msg, e)
    }
}
